paymentsheet/src/main/java/com/stripe/android/customersheet/data/CustomerSessionPaymentMethodDataSource.kt

```diff
@@ -1,30 +1,72 @@
 
+import com.stripe.android.core.injection.IOContext
 import com.stripe.android.model.PaymentMethod
 import com.stripe.android.model.PaymentMethodUpdateParams
+import com.stripe.android.payments.core.analytics.ErrorReporter
+import com.stripe.android.paymentsheet.repositories.CustomerRepository
+import kotlinx.coroutines.withContext
 import javax.inject.Inject
+import kotlin.coroutines.CoroutineContext
 
-internal class CustomerSessionPaymentMethodDataSource @Inject constructor() : CustomerSheetPaymentMethodDataSource {
+internal class CustomerSessionPaymentMethodDataSource @Inject constructor(
+    private val elementsSessionManager: CustomerSessionElementsSessionManager,
+    private val customerRepository: CustomerRepository,
+    private val errorReporter: ErrorReporter,
+    @IOContext private val workContext: CoroutineContext,
+) : CustomerSheetPaymentMethodDataSource {
     override suspend fun retrievePaymentMethods(): CustomerSheetDataResult<List<PaymentMethod>> {
-        throw NotImplementedError("Not implemented yet!")
+        return withContext(workContext) {
+            elementsSessionManager.fetchElementsSession().mapCatching { elementsSession ->
+                elementsSession.customer?.paymentMethods
+                    ?: throw IllegalStateException(
+                        "`CustomerSession` should contain payment methods in `elements/session` response!"
+                    )
+            }.toCustomerSheetDataResult()
+        }
     }
 
     override suspend fun updatePaymentMethod(
         paymentMethodId: String,
         params: PaymentMethodUpdateParams
     ): CustomerSheetDataResult<PaymentMethod> {
-        throw NotImplementedError("Not implemented yet!")
+        return withContext(workContext) {
+            elementsSessionManager.fetchCustomerSessionEphemeralKey().mapCatching { ephemeralKey ->
+                customerRepository.updatePaymentMethod(
+                    customerInfo = CustomerRepository.CustomerInfo(
+                        id = ephemeralKey.customerId,
+                        ephemeralKeySecret = ephemeralKey.ephemeralKey
+                    ),
+                    paymentMethodId = paymentMethodId,
+                    params = params,
+                ).getOrThrow()
+            }.toCustomerSheetDataResult()
+        }
     }
 
-    override suspend fun attachPaymentMethod(
-        paymentMethodId: String
-    ): CustomerSheetDataResult<PaymentMethod> {
-        throw NotImplementedError("Not implemented yet!")
+    override suspend fun attachPaymentMethod(paymentMethodId: String): CustomerSheetDataResult<PaymentMethod> {
+        errorReporter.report(
+            errorEvent = ErrorReporter.UnexpectedErrorEvent.CUSTOMER_SHEET_ATTACH_CALLED_WITH_CUSTOMER_SESSION,
+        )
+
+        return CustomerSheetDataResult.failure(
+            cause = IllegalStateException("'attach' is not supported for `CustomerSession`!"),
+            displayMessage = null,
+        )
     }
```

Comment:
In a later refactor, we should just handle the entire process of adding payment methods to a customer in isolation for `CustomerAdapter` and `CustomerSession`. This means moving the entire `IntentConfirmationHandler` logic in `CustomerSheetViewModel` into a separate source.

For now, returning a failure like `iOS` does and as an additional precaution, reporting an unexpected error. As we should never hit this method with `CustomerSession`.

Reply:
sounds good! How are we going to avoid calling this? Since in CustomerSheet, we shouldn't actually need to know which implementation of CustomerSheetPaymentMethodDataSource we are using, I'm not sure how we'd tell not to call attach or how that works

Reply:
What we would do remove the function from `PaymentMethodDataSource` and move it into `IntentDataSource`. We would probably rename the data source to `CustomerSheetConfirmationDataSource` or `CustomerSheetAttachDataSource` which can handle both attach with and without setup intents for `CustomerAdapter` and attach with ONLY setup intents for `CustomerSession`.